### PR TITLE
Flow tests with page objects

### LINKED ISSUE #10: Refactor flow tests to use page objects
#### As a developer I want my flow tests to be written using the object-oriented [page object pattern](http://martinfowler.com/bliki/PageObject.html).

_Assignees_: @kkajla12, @lsq921224, @ragurney

__Story__:

Now that we have gotten some experience writing flow tests with Capybara,
we've noticed that these tests a) can be brittle in that changing selectors
requires updating many tests and b) use a low-level interface that can make
reading and writing flow tests more cumbersome than we'd like.

Fortunately, there is a better way: using page objects. With page objects,
we can create objects to encapsulate pages, elements, and actions that can be
taken in our application. We can design an interface for interacting with our
application that maps directly to how our users would think about interacting
with the application. This interface makes the tests easier to read, write,
and modify, and reduces the possibility of writing flaky or invalid tests.

Working on this story will be a bit different than your typical stories. We've
created a starting point for you in the form of a branch in your repo called
`flowTestsWithPageObjects`. This branch contains flow tests for adding images,
deleting images, and viewing images associated with a tag (in the
`images_crud_test.rb` file). These tests are fully written, but the page
objects they use are not. Your job is to fill in the page objects to make the
tests pass against your application.

The first step is to rebase the `flowTestsWithPageObjects` branch on top of
your updated master branch. Note that if you already have an
`images_crud_test.rb` file, there will be a merge conflict, so you may want to
rename your file first, rebase, make the tests on that branch pass, and then
migrate any other tests that were in your file to the page object style before
deleting your old test file.

Finally, there is a video to help you get started. You should watch this video
and also take a look at the ae_page_objects readme which will be an extremely
helpful resource as you learn to write page objects. Both are linked in the
reference material below.

__Acceptance criteria__:
- [ ] The flow tests have been refactored to use page objects.

__Discussion topics__:
- [ ] What makes a good page object interface?

__Reference material__:
- https://github.com/appfolio/ae_page_objects
- https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project/flow-tests-with-page-objects

__Dependencies__:
- Delete Images
- Images by Tag

### PR DESCRIPTION
**Pull request by RitamBanerjee**
Closes #10 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: app/views/images/index.html.erb [hunk 1]
```diff
@@ -7,8 +7,8 @@
 
 <% if @images.size > 0 %>
   <% @images.each do |image| %>
-        <p>
-        <a href="<%= image_url(image.id)%>"><%= image.title %>  </a>
+        <p class='image-container'>
+        <a class='image-title' href="<%= image_url(image.id)%>"><%= image.title %>  </a>
```
Comment (ragurney) [inline app/views/images/index.html.erb]: Where does the `image_url` method come from?

File: app/views/images/show.html.erb [hunk 2]
```diff
@@ -5,7 +5,7 @@
 </p>
 
 <p>
-  <img src='<%=@image.url%>'>
+  <img class='show-image' src='<%=@image.url%>'>
```
Comment (ragurney) [inline app/views/images/show.html.erb]: What is this class used for? If it's only for tests/javascript then we usually prefix those kinds of classes with `js-`. This is to let other developers know that it's only used for those purposes, and also makes it so that those kinds of utility classes are not coupled with style class names, which may change later. So this would be `js-show-image`.

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -31,8 +31,6 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     assert_difference 'Image.count', 0 do
       post images_url, params: { image: { title: 'Test Image', url: 'http://i.imgur.com/qkdpN.jp', tag_list: 'tag1, tag2, tag3' } }
     end
-
-    assert_redirected_to new_image_url
```
Comment (ragurney) [inline test/controllers/images_controller_test.rb]: Could we `assert` something here about the html we expect to get back?  Perhaps `assert_select` would be useful here?

File: test/page_objects/document.rb [hunk 4]
```diff
@@ -1,7 +1,8 @@
 module PageObjects
   class Document < AePageObjects::Document
+
```
Comment (ragurney) [inline test/page_objects/document.rb]: Super nitpicky, but usually there is not a space between the class declaration and the first method of the class.

File: test/page_objects/images/index_page.rb [hunk 5]
```diff
@@ -15,7 +15,8 @@ def showing_image?(url:, tags: nil)
       end
 
       def clear_tag_filter!
-        #TODO
+        IndexPage.visit()
```
Comment (ragurney) [inline test/page_objects/images/index_page.rb]: Instead of a call to `visit`, could this be changed to an action a user could actually do to clear the tags? (i.e. clicking on a link that would take them to the index page).

File: app/views/images/show.html.erb [hunk 6]
```diff
@@ -10,6 +10,9 @@
 </p>
 
 <ul>
+  <%= link_to 'Destroy', image_path(@image),
```
Comment (ragurney) [inline app/views/images/show.html.erb]: Why was this added in this particular commit?

File: config/initializers/simple_form.rb [hunk 7]
```diff
@@ -0,0 +1,169 @@
+# Use this setup block to configure all options available in SimpleForm.
```
Comment (ragurney) [inline config/initializers/simple_form.rb]: What are these files for? Did you need them specifically for the tags test?

File: test/page_objects/images/image_card.rb [hunk 8]
```diff
@@ -14,11 +14,13 @@ def url
...
 
       def tags
-        #TODO
+        tags = node.all('.img-tag')
+        tags.map(&:text)
       end
 
       def click_tag!(tag_name)
-        #TODO
+        find("button", text: "#{tag_name}").click
```
Comment (ragurney) [inline test/page_objects/images/image_card.rb]: 👍

File: test/page_objects/images/image_card.rb [hunk 9]
```diff
@@ -19,7 +20,7 @@ def tags
       end
 
       def click_tag!(tag_name)
-        find("button", text: "#{tag_name}").click
+        find('button', text: tag_name.to_s).click
```
Comment (ragurney) [inline test/page_objects/images/image_card.rb]: Hm, do we even need the `to_s` call in this case?

File: config/initializers/simple_form.rb [hunk 10]
```diff
@@ -6,7 +6,7 @@
   # stack. The options given below are used to wrap the
```
Comment (ragurney) [inline config/initializers/simple_form.rb]: Good style fixes in this commit, however in general, I try to avoid style refactor commits like this. Try to catch these kinds of style errors in their respective commits.

File: app/views/images/show.html.erb (hunk 6 above)
Comment (RitamBanerjee) [reply app/views/images/show.html.erb]: my bad, this should be in the previous commit
